src/NodeVisitor/NamespaceStmtCreator.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the humbug/php-scoper package.
+ *
+ * Copyright (c) 2017 Théo FIDRY <theo.fidry@gmail.com>,
+ *                    Pádraic Brady <padraic.brady@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Humbug\PhpScoper\NodeVisitor;
+
+use Humbug\PhpScoper\NodeVisitor\Collection\NamespaceStmtCollection;
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\ClassLike;
+use PhpParser\Node\Stmt\Interface_;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\NodeVisitorAbstract;
+
+/**
+ * Creates a namespace for whitelisted classes and interfaces which belong to the global namespace.
+ *
+ * ```
+ * class AppKernel {}
+ * ```
+ *
+ * =>
+ *
+ * ```
+ * namespace Humbug;
+ *
+ * class AppKernel {}
+ * ```
+ */
+final class NamespaceStmtCreator extends NodeVisitorAbstract
+{
+    private $prefix;
+    private $hasWhitelistedNode;
+    private $namespaceStatements;
+    private $globalWhitelister;
+
+    /**
+     * @param string                  $prefix              Global prefix to apply.
+     * @param NamespaceStmtCollection $namespaceStatements List of statements in the current scope.
+     * @param callable                $globalWhitelister   List of whitelisted nodes.
+     */
+    public function __construct(
+        string $prefix,
+        NamespaceStmtCollection $namespaceStatements,
+        callable $globalWhitelister
+    ) {
+        $this->prefix = $prefix;
+        $this->namespaceStatements = $namespaceStatements;
+        $this->globalWhitelister = $globalWhitelister;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function beforeTraverse(array $nodes)
+    {
+        $classes = array_filter($nodes, [$this, 'isWhitelistableNode']);
+        if (empty($classes)) {
```

Comment:
as the returned type is an array I think [] !== $nodes or with a count is more expressive here